### PR TITLE
boot, overlord/devicestate: list trusted and managed assets upfront

### PR DESCRIPTION
**Pull request by bboozzoo**
Instead of doing a lazy listing of trusted and managed assets, we can do so
upfront when creating an update observer.

This is a follow up to #9440.

### CODE REVIEW

**Comment by pedronis** [comment]
@bboozzoo can't we do the same for install?

**Comment by bboozzoo** [comment]
We can, but I was a bit on a fence about this as it felt a bit off to pass a location that has not been mounted yet as ubuntu-boot root and only gets mounted once install.Run() runs. However, the mount point is known upfront as well and none of the bootloader calls we execute actually use rootdir. Maybe we could add a minimal info trusted bootloader interface, one that is documented to not use require rootdir to be set.

**Comment by pedronis** [comment]
> We can, but I was a bit on a fence about this as it felt a bit off to pass a location that has not been mounted yet as ubuntu-boot root and only gets mounted once install.Run() runs. However, the mount point is known upfront as well and none of the bootloader calls we execute actually use rootdir. Maybe we could add a minimal info trusted bootloader interface, one that is documented to not use require rootdir to be set.

we need to think it through, I don't see a big win in this if we continue doing it in both styles, differently

**Comment by pedronis** [comment]
@bboozzoo   so I looked at the code and the bootloaders seems they will panic if rootdir is "" and is used, so we could pass that in the Install case for now with a TODO to clarify and delineate the usage of bootloaders with rootdir ""

File: boot/assets.go [hunk 1]
```diff
@@ -376,20 +398,34 @@ type TrustedAssetsUpdateObserver struct {
...
+	return trustedAssets, managedAssets, nil
+}
+
+func findMaybeTrustedBootloaderAndAssets(rootDir string, opts *bootloader.Options) (foundBl bootloader.Bootloader, trustedAssets, managedAssets []string, err error) {
+	foundBl, err = bootloader.Find(rootDir, opts)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("cannot find bootloader: %v", err)
+	}
+	trustedAssets, managedAssets, err = trustedAndManagedAssetsOfBootloader(foundBl)
+	return foundBl, trustedAssets, managedAssets, err
```
Comment (pedronis) [inline boot/assets.go:419]: I don't think managedAssets is used now by any of the callers,  this means it probably shouldn't be returned

File: boot/assets.go [hunk 2]
```diff
@@ -341,27 +334,56 @@ func (o *TrustedAssetsInstallObserver) ChosenEncryptionKey(key secboot.Encryptio
...
+
+	// and the recovery bootloader
+	seedBl, seedTrusted, seedManaged, err := gadgetMaybeTrustedBootloaderAndAssets(gadgetDir, InitramfsUbuntuSeedDir,
+		&bootloader.Options{
+			Role: bootloader.RoleRecovery,
+		})
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (pedronis) [inline boot/assets.go:364]: if the various *Assets are empty (also based on trackTrustedAssets) we should return ErrObserverNotApplicable ?

File: boot/assets.go [hunk 3]
```diff
@@ -292,21 +290,16 @@ func (o *TrustedAssetsInstallObserver) ObserveExistingTrustedRecoveryAssets(reco
 		// assets do not need tracking, nothing to do
 		return nil
 	}
-	bl, err := bootloader.Find(recoveryRootDir, &bootloader.Options{
+	bl, trustedAssets, _, err := findMaybeTrustedBootloaderAndAssets(recoveryRootDir, &bootloader.Options{
```
Comment (pedronis) [inline boot/assets.go:293]: if we moved this to *ForModel already, we could also apply the logic that if all the assets are empty and useEncryption is false we could return ErrObserverNotApplicable

**Review by pedronis — COMMENTED** [review]
some comments about maybe possible further streamlining

File: boot/assets.go [hunk 4]
```diff
@@ -182,12 +182,49 @@ func TrustedAssetsInstallObserverForModel(model *asserts.Model, gadgetDir string
...
+	}
+	if !useEncryption {
+		// we do not care about trusted assets when not encrypting data
+		// partition
+		runTrusted = nil
+		seedTrusted = nil
+	}
+	hasManaged := len(runManaged) > 0
+	hasTrusted := len(runTrusted) > 0 || len(seedTrusted) > 0
+	if !hasManaged && !hasTrusted {
```
Comment (pedronis) [inline boot/assets.go:211]: maybe it relates to your wondering today, but given this is also used to track the actual key, I think we should also check && !useEncryption here?

**Review by pedronis — COMMENTED** [review]
thanks, one comment

File: boot/assets.go (hunk 4 above)
Comment (bboozzoo) [reply boot/assets.go:211]: I think it's fine to add it for now. With #9476 there's no more SetEncryptionKey.

Comment (bboozzoo) [reply boot/assets.go]: Pushed

**Review by pedronis — APPROVED** [review]
thank you

**Review by cmatsuoka — APPROVED** [review]
Thanks Maciek, lgtm.

**Comment by bboozzoo** [comment]
@mvo5 this one can land too. The failures are in tests/main/searching.

**Pull request merged**